[Mobile] - Cover block - Media editing

geriux opened the pull request:
`Gutenberg Mobile PR` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/2403
`WordPress iOS PR` -> https://github.com/wordpress-mobile/WordPress-iOS/pull/14420
`WordPress Android PR` -> https://github.com/wordpress-mobile/WordPress-Android/pull/12356

## Description
This PR adds [Media editing functionality](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2388) to the Cover block. With the [recent changes and refactor](https://github.com/WordPress/gutenberg/pull/22788) we can reuse the same functionality as in Image and Media & Text block.

Since Cover has some overlay views, I had to extract the media editing button to be able to render it on top, so I created a new component `ImageEditingButton` which is the one that the `Image` component uses but now can be imported from other components for more customization.

It also adds a new prop `pickerOptions` to the `MediaEdit` component to pass extra options for the picker.

## How has this been tested?

<details><summary>Steps to test</summary>
---

**Edit an image**

- Open the app with metro running
- Add a Cover block
- Tap on the block and upload a picture
- Once the image has been uploaded tap on it
- **Expect** the image to be selected and the media button option to be visible at the top right
- Tap on the media editing button
- **Expect** to see the options
- Tap on **Edit**
- **Expect** to see the media editor
- Edit the image and tap on done
- **Expect** to see the edited picture uploaded

**Replace an image**

- Open the app with metro running
- Add a Cover block
- Tap on the block and upload a picture
- Once the image has been uploaded tap on it
- **Expect** the image to be selected and the media button option to be visible at the top right
- Tap on the media editing button
- **Expect** to see the options
- Tap on **Replace**
- **Expect** to see the media picker
- Upload a new picture
- **Expect** to see the new picture uploaded

**Full screen preview**

- Open the app with metro running
- Add a Cover block
- Tap on the block and upload a picture
- Once the image has been uploaded tap on it
- **Expect** the image to be selected
- Tap on the image again
- **Expect** to see the image in full screen

</details>

## Screenshots 

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/87537190-89f4b380-c69a-11ea-8845-ce0954e1c569.gif" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/87537233-9842cf80-c69a-11ea-8d8a-a58ecceafa1a.gif" width="250" /></kbd>

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Commit:
Mobile - Cover block - Media editing

Commit:
Mobile - Cover block - Fix upload placeholder

geriux commented:
Hey @iamthomasbishop 👋 This is also ready for a design review whenever you can. Thank you!

mkevins commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -145,6 +168,10 @@ const Cover = ( {
 			requestImageUploadCancelDialog( id );
 		} else if ( shouldShowFailure ) {
 			requestImageFailedRetryDialog( id );
+		} else if ( backgroundType === MEDIA_TYPE_IMAGE && isMediaSelected ) {
+			requestImageFullscreenPreview( url );
+		} else if ( backgroundType === MEDIA_TYPE_IMAGE ) {
+			setMediaSelected( true );
```
Minor nit: I wonder if it'd be more readable to only test the `backgroundType` once, e.g.:

```suggestion
		} else if ( backgroundType === MEDIA_TYPE_IMAGE ) {
			if ( isMediaSelected ) {
				requestImageFullscreenPreview( url );
			} else {
				setMediaSelected( true );
			}
```

geriux replied:
Thanks for the suggestion Matt! I'll wait for the design feedback to see if this logic needs to change but if it doesn't I'll use that =)

mkevins commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -55,7 +57,8 @@ const ImageComponent = ( {
 	url,
 	width: imageWidth,
-	focalPoint,
+	focalPoint: focalPointValues,
+	withFocalPoint = false,
```
I might be missing some context on this, but I was wondering if we could make this interface smaller. Do we need to differentiate between `withFocalPoint` flag being false and `focalPoint` not being passed? In any case, if it's necessary, perhaps we could call `withFocalPoint` something else.. maybe `hasFocalPoint`, or even `isCropped` if that's what it really represents?

geriux replied:
Thanks for bringing it up! I'll give it a thought, I agree that we could make it smaller so it'd be ideal just to have `focalpoint` and that's it.

mkevins reviewed (commented):
I tested this on Android (via Pixel 3a) using the steps provided and everything is working as described! I also tried leaving the editor while the upload for the newly edit image was completing, and reopening the editor shows the correct image (url is the remote one). Nice work @geriux :tada: . Overall, code looks good, I just left a few suggestions.

iamthomasbishop commented:
@geriux Looks mostly good! The only thing that stands out to me is that you have to tap twice to select the background image before seeing the edit button. Would it be possible to instead show the edit button _when the parent is selected_? I think this might feel more intuitive.

geriux commented:
> @geriux Looks mostly good! The only thing that stands out to me is that you have to tap twice to select the background image before seeing the edit button. Would it be possible to instead show the edit button _when the parent is selected_? I think this might feel more intuitive.

Thanks for the feedback! So the thing is the Media edit button is included with the Image component, so its z-index only affects this component. In Cover, we have the overlay so if we make the button to show when the block is selected it'd show up like this:

<kbd><img src="https://user-images.githubusercontent.com/4885740/87158255-bcc83180-c2bf-11ea-9126-37b424852d61.png" width=200 /></kbd>

One workaround I tried is to hide the overlay when it's selected, and only show it when the block options are shown for example:

<kbd><img src="https://user-images.githubusercontent.com/4885740/87158466-13357000-c2c0-11ea-990b-54c10bed7998.gif" width=200 /></kbd>

But I'm not sure if it'd be confusing to make those opacity changes, what do you think?

Commit:
Mobile - Cover media editing - Image edit button

Commit:
Merge branch 'master' into rnmobile/cover-media-editing

# Conflicts:
#	packages/block-library/src/cover/edit.native.js
#	packages/block-library/src/cover/style.native.scss

geriux commented:
Hey @iamthomasbishop I managed to move the button 🎉 

<kbd><img src="https://user-images.githubusercontent.com/4885740/87456307-02119980-c607-11ea-9f93-2e1041bdcaf0.gif" /></kbd>

I'll work on getting some builds for you to test soon.

iamthomasbishop commented:
@geriux That's great news! 🥳  Will test when a new build is available!

Commit:
Mobile - Image editing button - Check for url before rendering

geriux commented:
Hey @iamthomasbishop builds are ready =) Thanks!

geriux commented:
Hey @mkevins 👋 this is ready for another review, thanks!

iamthomasbishop commented:
@geriux So glad you were able to move the button upward, it feels a lot more natural. While testing the latest build, I jotted down some additional feedback. Think it's pretty damn close!

- Now that we have the `Clear Media` action in the block settings sheet, do you think it'd make sense to also add it as an action in the Action/BottomSheet that right now contains `Edit` and `Replace`?
- Do you think it'd make sense to give that action/BottomSheet a title/description that is labeled `Media options`? I *think* we're getting this from the native side, so maybe we'd apply this across the board. We could get more specific with the label depending on its media type ( `Image options`, `Video options`, etc) but I think a more generic label like `Media options` would probably be fine if we want to keep it simple.
- Do you think it'd make sense to show a button in place of the edit button when I've cleared the media (either from the settings sheet or this action/bottom sheet)? ([example](https://cloudup.com/cfrOjoZe29v) — note: in this example, I uploaded an image bg, added a gradient overlay, and then cleared the media. I know we'll likely add more media attachment options in the settings sheet, but on top of the cell here seemed like the most obvious place to reach in this scenario)

iamthomasbishop commented:
@geriux Ah, I just noticed one more thing: the image selection outline (the blue inset border) is covered by the overlay background — is there any way to set the border to the outside of the image boundary or when selected, change the inset position of the overlay to account for this? [Here's an example](https://cloudup.com/cVzkpUwwvww), with and without an overlay.

mkevins commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -330,27 +348,45 @@ const Cover = ( {
+						isUploadFailed={ didUploadFail }
+						isUploadInProgress={ isUploadInProgress }
+						onSelectMediaUploadOption={ onSelectMedia }
+						openMediaOptions={ openMediaOptionsRef?.current }
```
I don't think we need the optional chaining for `openMediaOptionsRef`, it will always exist (from `useRef`). Wdyt?

geriux replied:
You're right! I'll remove the `?.` thanks!

mkevins commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -330,27 +348,45 @@ const Cover = ( {
 		<View style={ styles.backgroundContainer }>
 			{ controls }
 
+			{ openMediaOptionsRef?.current && (
```
Same here

mkevins commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -330,27 +348,45 @@ const Cover = ( {
 				onSelect={ onSelectMedia }
-				render={ renderBackground }
+				render={ ( { open, getMediaOptions } ) => {
+					openMediaOptionsRef.current = open;
```
If we are keeping `openMediaOptionsRef` in as a ref, do we still need to pass `open` to `renderBackground`?

geriux replied:
We don't :D I'll update the rest of the code so it uses the ref instead of passing it as a parameter.

mkevins commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -256,7 +260,9 @@ const Cover = ( {
-				{ isParentSelected && toolbarControls( openMediaOptions ) }
+				{ isParentSelected &&
+					backgroundType === VIDEO_BACKGROUND_TYPE &&
+					toolbarControls( openMediaOptions ) }
```
Can we use the ref.current here?

mkevins commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -284,12 +290,24 @@ const Cover = ( {
+							isUploadFailed={ didUploadFail }
+							isUploadInProgress={ isUploadInProgress }
+							onSelectMediaUploadOption={ onSelectMedia }
+							openMediaOptions={ openMediaOptions }
```
Same, can we use the ref.current here?

mkevins commented on packages/components/src/mobile/image/image-editing-button.native.js:
```diff
@@ -0,0 +1,52 @@
+import styles from './style.scss';
+
+const ImageEditingButton = ( {
+	isSelected,
+	isUploadFailed,
+	isUploadInProgress,
```
Since we only use these props to decide _not_ to render the component, maybe this could be simplified by conditioning in the parent. Wdyt?

geriux replied:
Ohh I like that! Thanks for the suggestion, I'll update it =)

mkevins reviewed (commented):
I tested a few flows, and it's looking pretty good with the new button behavior. I left a few minor comments, but overall it's looking good. :+1: 

I'll defer fully testing (e.g. upload failure / retry / progress, etc.) until after the other design issue(s) are resolved. Btw, the gallery image has inset border in a separate layer, so it's not being dimmed by the overlay, in case that helps with ideas.

Commit:
Mobile - Cover - Show selected border with the overlay layer

Commit:
Merge branch 'master' into rnmobile/cover-media-editing

geriux commented:
Thank you for the feedback @iamthomasbishop !

> * Now that we have the `Clear Media` action in the block settings sheet, do you think it'd make sense to also add it as an action in the Action/BottomSheet that right now contains `Edit` and `Replace`?

We are adding `Remove` for the gallery, I think this would work for this case too. What do you think? I'm attaching a gif below.

> * Do you think it'd make sense to give that action/BottomSheet a title/description that is labeled `Media options`? I _think_ we're getting this from the native side, so maybe we'd apply this across the board. We could get more specific with the label depending on its media type ( `Image options`, `Video options`, etc) but I think a more generic label like `Media options` would probably be fine if we want to keep it simple.

I think we should, I added it in the example below.

> * Do you think it'd make sense to show a button in place of the edit button when I've cleared the media (either from the settings sheet or this action/bottom sheet)? ([example](https://cloudup.com/cfrOjoZe29v) — note: in this example, I uploaded an image bg, added a gradient overlay, and then cleared the media. I know we'll likely add more media attachment options in the settings sheet, but on top of the cell here seemed like the most obvious place to reach in this scenario)

I think it'd make sense, I saw this in a mockup if I'm not mistaken, with all the Cover work with solid colors and now these changes with media editing we could easily add it.

> @geriux Ah, I just noticed one more thing: the image selection outline (the blue inset border) is covered by the overlay background — is there any way to set the border to the outside of the image boundary or when selected, change the inset position of the overlay to account for this? Here's an example, with and without an overlay.

Nice catch! I'll fix it.

Here's a gif with all the changes, what do you think?

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/87770752-b8e86200-c81f-11ea-87f3-742e780568e1.gif" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/87770757-bab22580-c81f-11ea-860a-75af5cd078fc.gif" width="250" /></kbd>

For Android, as you can see in the gif above the options have icons. Would you want to remove the icons as we did with the block options menu? e.g.

<kbd><img src="https://user-images.githubusercontent.com/4885740/87770869-dc131180-c81f-11ea-80e7-262942b0a59c.png" width="250" /></kbd>

Commit:
Mobile - Cover - Style improvements, Media editing support for onRemove

Commit:
Merge branch 'master' into rnmobile/cover-media-editing

# Conflicts:
#	packages/block-library/src/cover/edit.native.js
#	packages/block-library/src/cover/style.native.scss

Commit:
Mobile - Cover - Remove extra line in style

iamthomasbishop commented:
> We are adding Remove for the gallery, I think this would work for this case too. What do you think? I'm attaching a gif below.

@geriux I think the `Remove` action on the gallery block is doing something slightly different -- it's removing the image cell altogether. Whereas for the cover block, it's only explicitly removing the background image. So perhaps the label here should be `Clear media`, which would match our settings button?

> I think we should, I added it in the example below.

Looks good!

> I think it'd make sense, I saw this in a mockup if I'm not mistaken, with all the Cover work with solid colors and now these changes with media editing we could easily add it.

Looks like @antonis already added a [ticket](https://github.com/WordPress/gutenberg/pull/23878) to work on that separately, which I hadn't even seen earlier -- great minds think alike? 😉 

> Here's a gif with all the changes, what do you think?

Looking great!

> For Android, as you can see in the gif above the options have icons. Would you want to remove the icons as we did with the block options menu? e.g.

I think we can remove the icons, yes. This reminds me I need to write some clear guidelines for when/when not to use icons 😄

Commit:
Mobile - Media editing - Support for more picker options

geriux commented:
> @geriux I think the `Remove` action on the gallery block is doing something slightly different -- it's removing the image cell altogether. Whereas for the cover block, it's only explicitly removing the background image. So perhaps the label here should be `Clear media`, which would match our settings button?

Sounds good! I'll add `Clear media` for this one and `Remove` for gallery.

> I think we can remove the icons, yes. This reminds me I need to write some clear guidelines for when/when not to use icons 😄

Awesome, thanks!

Commit:
Mobile - Cover - Update clear media label

Commit:
Mobile - Media Edit - Check if optionSelected is defined

Commit:
Merge branch 'master' into rnmobile/cover-media-editing

geriux commented:
Hey @iamthomasbishop I've just made new builds with the latest feedback, just in case you want to give another quick check.

@mkevins This is ready for another review 🙇 Thanks!

mkevins commented:
This is looking good. I tested lots of flows on Android, including media upload failure, retry, cancel, as well as the editing flows like editing, changing media, removing media, etc. And all of that is working as expected. Also, the inset border is not covered by the overlay :tada:.

The only potential blocker is with image preview: it seems that tapping the cover block without an image (but with a color) still tries to display the preview, even though there is no media. This results in a blank screen:

<img src="https://user-images.githubusercontent.com/8507675/88150346-9f428280-cc44-11ea-9519-a2e783edd6db.gif" width="360">

Also, since it automattically returns from the blank preview screen after a few seconds (with a toast), sometimes when you tap the back button, it leaves the editor instead. If you tap early enough, it just goes back to the editor, but if you tap just as the preview goes away on its own, the post is closed.

Also, one thing to note, (though it may not be directly related this PR), it seems after adding color, there is no way to change the background to media.. or maybe I just couldn't find the option. Is this known / expected?

Commit:
Mobile - Cover - Don't show the full screen modal preview if there is not a url set

Commit:
Mobile - Update release notes

geriux commented:
> This is looking good. I tested lots of flows on Android, including media upload failure, retry, cancel, as well as the editing flows like editing, changing media, removing media, etc. And all of that is working as expected. Also, the inset border is not covered by the overlay 🎉.

Awesome! Thanks for reviewing and testing those flows!
 
> The only potential blocker is with image preview: it seems that tapping the cover block without an image (but with a color) still tries to display the preview, even though there is no media. This results in a blank screen:

Nice catch! I [added this](https://github.com/WordPress/gutenberg/pull/23280/commits/02aaf3dbe3a969b51fcaaeb06a882774b632c824) to check if there's an URL set. Didn't realize the the `backgroundType` is set to `image` by default.

> Also, since it automattically returns from the blank preview screen after a few seconds (with a toast), sometimes when you tap the back button, it leaves the editor instead. If you tap early enough, it just goes back to the editor, but if you tap just as the preview goes away on its own, the post is closed.

I think this was due to passing an undefined URL to the preview so it should work as expected now.

> Also, one thing to note, (though it may not be directly related this PR), it seems after adding color, there is no way to change the background to media.. or maybe I just couldn't find the option. Is this known / expected?

It is known and it's being handled in [another PR](https://github.com/WordPress/gutenberg/pull/23878). Thanks for bringing it up though =D

Commit:
Merge branch 'master' into rnmobile/cover-media-editing

Commit:
Mobile - Update changelog

mkevins reviewed (approved):
Tested again after your latest changes, and it's fixed now. Nice work @geriux ! LGTM!

geriux commented:
> Tested again after your latest changes, and it's fixed now. Nice work @geriux ! LGTM!

Thank you so much for reviewing and testing this @mkevins!

Hey @iamthomasbishop 👋  Let me know if you want to review this again or if its good to merge. Thanks!

iamthomasbishop commented:
@geriux Looking great to me! I noticed one tiny issue but it's already being resolved in [another issue](https://github.com/WordPress/gutenberg/pull/23878#issuecomment-662855005), so I think this one is good to ship!
